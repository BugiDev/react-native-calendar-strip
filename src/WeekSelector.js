import React, { Component } from "react";
import PropTypes from "prop-types";
import { Image, TouchableOpacity } from "react-native";

import moment from "moment";

import styles from "./Calendar.style.js";

class WeekSelector extends Component {
  static propTypes = {
    controlDate: PropTypes.object,
    iconComponent: PropTypes.any,
    iconContainerStyle: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.number
    ]),
    iconInstanceStyle: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.number
    ]),
    iconStyle: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.number,
      PropTypes.array
    ]),
    imageSource: PropTypes.oneOfType([PropTypes.object, PropTypes.number]),
    onPress: PropTypes.func,
    weekStartDate: PropTypes.object,
    weekEndDate: PropTypes.object
  };

  shouldComponentUpdate(nextProps) {
    // Extract iconComponent since JSON.stringify fails on React component circular refs
    let _nextProps = Object.assign({}, nextProps);
    let _props = Object.assign({}, this.props);

    delete _nextProps.iconComponent;
    delete _props.iconComponent;

    return (
      JSON.stringify(_props) !== JSON.stringify(_nextProps) ||
      this.props.iconComponent !== nextProps.iconComponent
    );
  }

  isEnabled(controlDate, weekStartDate, weekEndDate) {
    if (controlDate) {
      return !moment(controlDate).isBetween(
        weekStartDate,
        weekEndDate,
        "day",
        "[]"
      );
    }
    return true;
  }

  render() {
    const {
      controlDate,
      iconContainerStyle,
      iconComponent,
      iconInstanceStyle,
      iconStyle,
      imageSource,
      onPress,
      weekEndDate,
      weekStartDate,
      size
    } = this.props;

    const enabled = this.isEnabled(controlDate, weekStartDate, weekEndDate);
    const opacity = { opacity: enabled ? 1 : 0 };

    let component;
    if (React.isValidElement(iconComponent)) {
      component = React.cloneElement(iconComponent, {
        style: [iconComponent.props.style, { opacity: opacity.opacity }]
      });
    } else if (Array.isArray(iconComponent)) {
      component = iconComponent;
    } else {
      let imageSize = { width: size, height: size };
      component = (
        <Image
          style={[
            styles.icon,
            imageSize,
            iconStyle,
            iconInstanceStyle,
            opacity
          ]}
          source={imageSource}
        />
      );
    }

    return (
      <TouchableOpacity
        style={[styles.iconContainer, iconContainerStyle]}
        onPress={onPress}
        disabled={!enabled}
      >
        {component}
      </TouchableOpacity>
    );
  }
}

export default WeekSelector;
