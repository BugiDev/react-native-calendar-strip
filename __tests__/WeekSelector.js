import React from "react";
import { Image, Text } from "react-native";

import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import moment from "moment";
import _ from "lodash";

import WeekSelector from "../src/WeekSelector";
import defaultStyles from "../src/Calendar.style.js";

const genDatesForWeek = startDate => {
  const day = moment(startDate);
  return _.map(_.range(7), i => day.clone().add(i, "days"));
};

describe("WeekSelector Component", () => {
  it("should render without issues", () => {
    const component = shallow(
      <WeekSelector imageSource={require("./img/left-arrow-black.png")} />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should have onPress event", () => {
    const onPress = jest.fn();
    const component = shallow(
      <WeekSelector
        imageSource={require("./img/left-arrow-black.png")}
        onPress={onPress}
      />
    );

    component.simulate("press");
    expect(onPress).toHaveBeenCalled();
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should have the default container style", () => {
    const component = shallow(
      <WeekSelector imageSource={require("./img/left-arrow-black.png")} />
    );

    const styles = component.props().style;

    expect(_.filter(styles)).toHaveLength(1);
    expect(styles[0]).toEqual(defaultStyles.iconContainer);
  });

  it("should have a custom container style of color red", () => {
    const component = shallow(
      <WeekSelector
        iconContainerStyle={{ color: "red" }}
        imageSource={require("./img/left-arrow-black.png")}
      />
    );

    const styles = component.props().style;

    expect(_.filter(styles)).toHaveLength(2);
    expect(styles[styles.length - 1].color).toBe("red");
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should be enabled", () => {
    const datesForWeek = genDatesForWeek("2018-01-10");

    const component = shallow(
      <WeekSelector
        imageSource={require("./img/left-arrow-black.png")}
        controlDate={moment("2018-01-08")}
        weekEndDate={datesForWeek[datesForWeek.length - 1]}
        weekStartDate={datesForWeek[0]}
      />
    );
    expect(component.props().disabled).toBe(false);
  });

  it("should be disabled", () => {
    const datesForWeek = genDatesForWeek("2018-01-10");

    const component = shallow(
      <WeekSelector
        imageSource={require("./img/left-arrow-black.png")}
        controlDate={moment("2018-01-12")}
        weekEndDate={datesForWeek[datesForWeek.length - 1]}
        weekStartDate={datesForWeek[0]}
      />
    );
    expect(component.props().disabled).toBe(true);
    expect(toJson(component)).toMatchSnapshot();
  });
});

describe("WeekSelector Image Component ", () => {
  it("should have render an image", () => {
    const component = shallow(
      <WeekSelector imageSource={require("./img/left-arrow-black.png")} />
    );

    const imageComponent = component.find(Image);
    expect(imageComponent).toHaveLength(1);
  });

  it("should show the default styles", () => {
    const component = shallow(
      <WeekSelector imageSource={require("./img/left-arrow-black.png")} />
    );

    const imageComponent = component.find(Image);
    const styles = imageComponent.props().style;

    // REM the opacity is applied by default
    expect(_.filter(styles)).toHaveLength(2);
    expect(styles[0]).toEqual(defaultStyles.icon);
  });

  it("should have custom icon style of backgroundColor red", () => {
    const component = shallow(
      <WeekSelector
        iconStyle={{ backgroundColor: "red" }}
        imageSource={require("./img/left-arrow-black.png")}
      />
    );

    const imageComponent = component.find(Image);
    const styles = imageComponent.props().style;
    const finalStyle = _.findLast(styles, "backgroundColor");

    // REM the opacity is applied by default
    expect(_.filter(styles)).toHaveLength(3);
    expect(finalStyle).not.toBeNull();
    expect(finalStyle.backgroundColor).toEqual("red");
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should have custom icon instance style of backgroundColor blue", () => {
    const component = shallow(
      <WeekSelector
        iconStyle={{ backgroundColor: "red" }}
        iconInstanceStyle={{ backgroundColor: "blue" }}
        imageSource={require("./img/left-arrow-black.png")}
      />
    );

    const imageComponent = component.find(Image);
    const styles = imageComponent.props().style;
    const finalStyle = _.findLast(styles, "backgroundColor");

    // REM the opacity is applied by default
    expect(_.filter(styles)).toHaveLength(4);
    expect(finalStyle).not.toBeNull();
    expect(finalStyle.backgroundColor).toEqual("blue");
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should be enabled with opacity 1", () => {
    const datesForWeek = genDatesForWeek("2018-01-10");

    const component = shallow(
      <WeekSelector
        imageSource={require("./img/left-arrow-black.png")}
        controlDate={moment("2018-01-08")}
        weekEndDate={datesForWeek[datesForWeek.length - 1]}
        weekStartDate={datesForWeek[0]}
      />
    );

    const imageComponent = component.find(Image);
    const styles = imageComponent.props().style;
    const finalStyle = _.findLast(styles, "opacity");

    expect(finalStyle).not.toBeNull();
    expect(finalStyle.opacity).toBe(1);
  });

  it("should be disabled with opacity 0", () => {
    const datesForWeek = genDatesForWeek("2018-01-10");

    const component = shallow(
      <WeekSelector
        imageSource={require("./img/left-arrow-black.png")}
        controlDate={moment("2018-01-12")}
        weekEndDate={datesForWeek[datesForWeek.length - 1]}
        weekStartDate={datesForWeek[0]}
      />
    );
    const imageComponent = component.find(Image);
    const styles = imageComponent.props().style;
    const finalStyle = _.findLast(styles, ["opacity", 0]);

    expect(finalStyle).not.toBeNull();
    expect(finalStyle.opacity).toBe(0);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render a custom component", () => {
    const component = shallow(
      <WeekSelector iconComponent={<Text>GO BACK</Text>} />
    );

    const customComponent = component.find(Text);
    expect(customComponent).toHaveLength(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render a custom component that is disabled", () => {
    const datesForWeek = genDatesForWeek("2018-01-10");

    const component = shallow(
      <WeekSelector
        iconComponent={<Text style={{ color: "red" }}>GO BACK</Text>}
        controlDate={moment("2018-01-12")}
        weekEndDate={datesForWeek[datesForWeek.length - 1]}
        weekStartDate={datesForWeek[0]}
      />
    );

    const customComponent = component.find(Text);
    const styles = customComponent.props().style;
    const finalStyle = _.findLast(styles, ["opacity", 0]);

    expect(finalStyle).not.toBeNull();
    expect(finalStyle.opacity).toBe(0);
    expect(toJson(component)).toMatchSnapshot();
  });
});
