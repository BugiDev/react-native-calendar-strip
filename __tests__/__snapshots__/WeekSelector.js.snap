// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekSelector Component should be disabled 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={true}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        undefined,
        undefined,
        Object {
          "opacity": 0,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Component should have a custom container style of color red 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        undefined,
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Component should have onPress event 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        undefined,
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Component should render without issues 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        undefined,
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Image Component  should be disabled with opacity 0 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={true}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        undefined,
        undefined,
        Object {
          "opacity": 0,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Image Component  should have custom icon instance style of backgroundColor blue 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        Object {
          "backgroundColor": "red",
        },
        Object {
          "backgroundColor": "blue",
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Image Component  should have custom icon style of backgroundColor red 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "height": 20,
          "resizeMode": "contain",
          "width": 20,
        },
        Object {
          "backgroundColor": "red",
        },
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</TouchableOpacity>
`;

exports[`WeekSelector Image Component  should render a custom component 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={false}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        undefined,
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    GO BACK
  </Text>
</TouchableOpacity>
`;

exports[`WeekSelector Image Component  should render a custom component that is disabled 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  disabled={true}
  focusedOpacity={0.7}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "red",
        },
        Object {
          "opacity": 0,
        },
      ]
    }
  >
    GO BACK
  </Text>
</TouchableOpacity>
`;
